### ISSUE TITLE
`--no-config` with `--rulesdir` not works as expected

### ISSUE THREAD

**Issue by therynamo** [coderaiser/putout#138]
Hi @coderaiser , I see you've spent a lot of time on documentation and it is very much appreciated. 

I have been working for a couple of days trying to get a `user plugin` that I want to reference locally and I just can't seem to get it working. I have read through the docs a number of times, read through the code, but I still don't think I'm quite getting it. 

I'll describe what I'm trying to do - and maybe you can help shed some light on how I might accomplish this. 

---

I'd like to leverage `putout` as a codemod tool, in place of `jscodeshift`. I find the prospect of using `PutoutScript` to write codemods very interesting. The succinctness of the code to `replace` something like an `object property accessor` with `optional chaining` being so simple (e.g. `'__a && __a.__b': '__a?.__b'`). The requisite code to do something like this in `jscodeshift` would definitely be more complicated. 

I do only want to use this tool for `codemods` though. I had the desire to expose the `putout` library to run any number of `codemods` that I would put in this said repo. 

For example:

```
my-project/
|
|__ src/
|      |__codemods/
|      |    |__ putout-plugin-my-custom-codemod/
|      |         |__index.js
|      |    |__ putout-plugin-my-other-custom-codemod/
|      |         |__index.js
|      |__ index.js
|__ index.js
|__ package.json
|__ bob.ts
```

Lets say `./src/codemods/putout-plugin-my-custom-codemod/index.js` looks like this:

```js
module.exports.report = () => 'use optional chaining';
module.exports.replace = () => ({
  '__a && __a.__b': '__a?.__b',
});
```

And I have a file named `./bob.ts` that looks like this:

```ts
const wow = {};

const bobby = wow && wow.blue;

console.log({ bobby, wow })
```

I am running the `putout` CLI like so:

```shell
# remembering that the file `bob.ts` exists for the below example

$ putout --rulesdir src/codemods bob.ts --no-config --fix --plugins 'my-custom-codemod' --format dump
```

My understanding of what the above should do is the following:

- `putout` - runs the CLI binary 
- `--rulesdir src/codemods` - points `putout` to `./src/codemods` as the reference for where to find `plugins` [according to these docs](https://github.com/coderaiser/putout#rulesdir)
- `bob.ts` is the file that we're running the `putout` command on (with only the `my-custom-codemod` plugin enabled)
- `--no-config` - as per the `--no-config` issue - https://github.com/coderaiser/putout/issues/137#issuecomment-1500398708 - this tells `putout` there is `zero` configuration
- `--fix`: Tells `putout` to `write` to disk if there are applicable fixes
- `--plugins 'my-custom-codemod'` - tells `putout` what plugins to run as a part of this command. In this case it is only one plugin `my-custom-codemod`
- `--format dump` - simply `dump`s the output into terminal

Is any of what I have just outlined inaccurate? 

I was able to verify that running the API directly - the output is transformed as expected. See this example:

```js
const customMod = require('./src/codemods/putout-plugin-my-custom-codemod/index.js');

const { code } = putout(`
const wow = {};

const bobby = wow && wow.blue;

console.log({ bobby, wow })
`, { plugins: [ ['my-custom-codemod', customMod] ] })

console.log(code);

// output
// `
// const wow = {};
//
// const bobby = wow?.blue;
//
// console.log({ bobby, wow })
//`
```

I want to take advantage of all the work you've done in the CLI to stream files with glob patterns, and apply `replace`ments/`transforms` to all files. But the functionality I want is demonstrated above with the isolation of the custom plugin that is referenced. 

Essentially I'm trying to run `1-n` number of plugins that are created and maintained in `./src/codemods` when using the `putout` cli. 

What is the best way to do this?

**Comment by therynamo** [comment]
So I've done some digging - and it looks like the `require.resolve` calls are not finding the `putout-plugin-my-custom-codemod/index.js' - the `path`s here end up being `undefined`. 

https://github.com/coderaiser/putout/blob/30c4fe92917331f7c53896f04291e0ef356aa882/packages/engine-loader/lib/load.js#L51-L65

Which in turn causes the plugin to not be found. 

Changing `path` to a hardcoded string of the path to the custom plugin (e.g. `$HOME/my-project/src/codemods/putout-plugin-mycustom-codemod/index.js`). Once I did this - the plugin loaded and ran successfully. 

Is there another check we could do here to verify a local path? I'm piecing as much of this as I can together by just reading the code - but you probably have a clear thought as to what we might be able to do here.

**Comment by therynamo** [comment]
Oof - I think this came down to user error. 

So - full disclosure - I'm writing this in `TypeScript` (by the way - this package wouldn't happen to have any Types would it 😄 ?) - and that means it was getting compiled to a `dist` directory. 

I had the path wrong to the `rulesdir` so it wasn't finding any rules - which technically the error told me. 

I'm wondering if there could be an error thrown (or warning) if `rulesdir` doesn't exist - or is empty/has no rules. I know you can't account for every scenario a user might find themselves in. But just knowing the path of the failed import would be useful in this situation I think. 

So TL;DR - I think everything is working as expected now. 

- example plugin is picked up when `rulesdir` is set properly
- codemod runs on `file` passed in (e.g. `bob.ts`) and writes it when `--fix` is run

This is exactly what I was trying to do. Although - the `message` doesn't show up when using `--format=dump`. Do you have any recommendations on a `format` in this situation?

**Comment by therynamo** [comment]
Okay - I think I've finally figured out why what I'm trying to accomplish is not working. 

https://github.com/coderaiser/putout/blob/30c4fe92917331f7c53896f04291e0ef356aa882/packages/putout/lib/parse-options/index.js#L138

[From the documentation](https://github.com/coderaiser/putout#-codemods)

> 🐊Putout supports codemodes in the similar to plugins way, just create a directory ~/.putout and put your plugins there. Here is example: [convert-tape-to-supertape](https://github.com/coderaiser/putout/tree/v12.0.0/codemods/convert-tape-to-supertape#install) and [this is example of work](https://github.com/coderaiser/putout/commit/ad02cebc344ce73cdee668cffc5078bf08830d52).

The key callout here being `~/.putout`. Right below this section - there is a callout for [`rulesdir`](https://github.com/coderaiser/putout#rulesdir)

> This way you can keep rules specific for your project and run them on each lint.

Which made it seem like the plugins would get loaded if I told `putout` where they were located. So I wouldn't have to utilize the `~/` home dir. 

When looking at `parse-options`

https://github.com/coderaiser/putout/blob/30c4fe92917331f7c53896f04291e0ef356aa882/packages/putout/lib/cli/get-options.js#L24-L33

The `--no-config` flag `ignores` the `rulesdir` flag - since the `rulesdir` flag isn't used until after short circuting. 

I guess maybe I'm not fully understanding the `--no-config` flag - or if its helping me try and accomplish what I'm trying to do. 

But when `parseOptions` is run, like when I'm debugging I can run it before the short circut, I find that the `custom-example` plugin gets clobbered here:

https://github.com/coderaiser/putout/blob/master/packages/putout/lib/parse-options/index.js#L60

Moving this line to the end of the `merge` allows for the `custom-example` plugin to get loaded in accordingly. This however comes with the additional rules that are all loaded by default. 

I know there is the disable all flag from the cli:

```
   --disable-all               disable all found rules (set baseline) and save them to '.putout.json' walking up parent directories
```

But this is limited to rules that apply to a given path. 

> This command will disable all rules that 🐊Putout can find right now and enable a single rule. All built-in rules made for good and highly suggested to be used, all of them are enabled in all my repositories, since they have auto fix.

> ☝️You can always disable what you don't need, so give it a try. You won't regret 🐊.


While I appreciate this sentiment in the documentation - I really am just looking for the usage of the `codemod` functionality at this time.

**Comment by therynamo** [comment]
I took a stab at what supporting this could look like:

```js

    let customPlugins = [];

    if (rulesdir) {
        customPlugins = readRules('./', rulesdir).plugins || [];
    }
   
    if (noConfig)
        return {
            ...maybeConfig,
            dir: dirname(name),
            plugins: [
                ...transformPlugins,
                ...maybeConfig.plugins,
                ...customPlugins,
            ],
        };
 
```

Where the key differences to note are:

```diff
- ...plugins
+ ...customPlugins
```

If we're loading custom plugins from a directory - we don't need the string names since those will be provided by `readRules`. They clash and cause the plugin to fail - so removing `...plugin` allows the `custom-example` plugin to run. 

And the other big addition being:

```diff
+   if (rulesdir) {
+        customPlugins = readRules('./', rulesdir).plugins || [];
+   }
```

This makes a couple of assumptions. First that `./` is the intended root path. `process.cwd()` may be more appropriate here (who is to say 🤷). Second is _the idea_ that when passing **both** `--no-config` and `--rulesdir`, you want to load custom plugins from that specified path (even though `--no-config` was passed). 

If that's okay - then this will work. If that's not the right solution for this problem - I'm open to ideas. But I do think support for this would be lovely.

**Comment by coderaiser** [comment]
So much information, I'll try to answer to some of them:
1. you can find types in branch [`@feature/typescript-support-for-plugins`](https://github.com/coderaiser/putout/tree/feature/typescript-support-for-plugins), add send pr  to [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) marking me in PR.
2. `--no-config` was made for tests mostly, it is much better to use `.putout.json` config disabling rules that you don't need (most of them you definitely need!), anyways it is possible to use `no-config` with `--rulesdir`, here is how it can look like:

```js
putout --no-config --rulesdir rules/ -f dump .madrun.mjs
```

**Comment by coderaiser** [comment]
Anyways, I will be glad if you implement features you need, if you still sure that this is the right thing to do. But let's start from https://github.com/coderaiser/putout/issues/137, and then proceed step-by-step.

**Comment by coderaiser** [comment]
Closed due to a long time of inactivity 🏝 Fee free to reopen when you have chance to work on this.
